cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -75,48 +75,55 @@ struct CountImpl : public ScalarAggregator {
 
   Status Finalize(KernelContext* ctx, Datum* out) override {
     const auto& state = checked_cast<const CountImpl&>(*ctx->state());
-    switch (state.options.count_mode) {
-      case CountOptions::COUNT_NON_NULL:
-        *out = Datum(state.non_nulls);
-        break;
-      case CountOptions::COUNT_NULL:
-        *out = Datum(state.nulls);
-        break;
-      default:
-        return Status::Invalid("Unknown CountOptions encountered");
+    if (state.options.skip_nulls) {
+      *out = Datum(state.non_nulls);
+    } else {
+      *out = Datum(state.nulls);
```

Comment:
`skip_nulls = true` matches `COUNT_NON_NULL`, the meaning is clear.

`skip_nulls = false` is a bit confusing. E.g., `count(list, skip_nulls=False)` should return the total length including nulls, but it actually returns only #nulls.
I'm okay to accept current code, just wondering if there's better way.

Reply:
If `list.length` is not available it (`count(list, skip_nulls=False) - count(list, skip_nulls=True)`) does mean executing `count` kernel twice.

Reply:
Yes. But for current kernel, you have to call it twice if you want the array length. So looks fair.

Reply:
Ok, I'm pro `count(list, skip_nulls=False) == length_of_list` then. Let's see what @jorisvandenbossche thinks.

Reply:
Actually if we do change to `count(list, skip_nulls=False) == length_of_list` option then we break with the [NaiveGroupBy comparison](https://github.com/apache/arrow/blob/ce2861713472818eea264957de4cc83d5a2c567c/cpp/src/arrow/compute/kernels/hash_aggregate_test.cc#L113). So I'd actually prefer `count(list, skip_nulls=False) == null_count`.
I suppose it's less intuitive but  it's more consistent.